service/test/dinstaller/software/proposal_test.rb

```diff
@@ -0,0 +1,175 @@
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require_relative "../../test_helper"
+require "dinstaller/software/proposal"
+require "dinstaller/config"
+
+describe DInstaller::Software::Proposal do
+  subject(:proposal) { described_class.new(logger: logger) }
+
+  let(:logger) { Logger.new($stdout) }
+  let(:destdir) { "/mnt" }
+  let(:result) { {} }
+  let(:last_error) { "" }
+  let(:solve_errors) { 0 }
+
+  before do
+    allow(Yast::Pkg).to receive(:SourceSaveAll)
+    allow(Yast::Packages).to receive(:Proposal).and_return(result)
+    allow(Yast::Pkg).to receive(:TargetFinish)
+    allow(Yast::Pkg).to receive(:TargetInitialize)
+    allow(Yast::Pkg).to receive(:TargetLoad)
+    allow(Yast::Installation).to receive(:destdir).and_return(destdir)
+    allow(Yast::Pkg).to receive(:LastError).and_return(last_error)
+    allow(Yast::Pkg).to receive(:PkgSolveErrors).and_return(solve_errors)
+    allow(Yast::Pkg).to receive(:SetSolverFlags)
+  end
+
+  describe "#calculate" do
+    it "initializes the packaging target" do
+      expect(Yast::Pkg).to receive(:TargetFinish)
+      expect(Yast::Pkg).to receive(:TargetInitialize).with(destdir)
+      expect(Yast::Pkg).to receive(:TargetLoad)
+      subject.calculate
+    end
+
+    it "makes a proposal" do
+      expect(Yast::Packages).to receive(:Proposal).and_return(result)
+      expect(Yast::Pkg).to receive(:PkgSolve)
+      subject.calculate
+    end
+
+    it "selects the language packages" do
+      expect(Yast::Pkg).to receive(:SetAdditionalLocales).with(["de_DE"])
+      subject.languages = ["de_DE"]
+      subject.calculate
+    end
+
+    it "returns true" do
+      expect(subject.calculate).to eq(true)
+    end
+
+    context "when a proposal is not possible or contain errors" do
+      let(:solve_errors) { 1 }
+
+      it "returns false" do
+        expect(subject.calculate).to eq(false)
+      end
+    end
+
+    context "when no errors were reported" do
+      it "does not register any error" do
+        subject.calculate
+        expect(subject.errors).to be_empty
+      end
+    end
+
+    context "when a blocking warning is reported" do
+      let(:result) do
+        { "warning_level" => :blocker, "warning" => "Could not install..." }
+      end
+
+      it "registers the corresponding validation error" do
+        subject.calculate
+        expect(subject.errors).to eq(
+          [DInstaller::ValidationError.new("Could not install...")]
+        )
+      end
+    end
+
+    context "when solver errors are reported" do
+      let(:last_error) { "Solving errors..." }
+      let(:solve_errors) { 5 }
+
+      it "returns something" do
```

Comment:
NP: I'd say "registers them as validation errors" or so. In fact, you're not testing the return value of #calculate here.